src/compiler/parser.ts

```diff
@@ -4840,10 +4866,10 @@ namespace ts {
                 factory.createConditionalExpression(
                     leftOperand,
                     questionToken,
-                    doOutsideOfContext(disallowInAndDecoratorContext, parseAssignmentExpressionOrHigher),
+                    doOutsideOfContext(disallowInAndDecoratorContext, () => parseAssignmentExpressionOrHigher(/*disallowReturnTypeInArrowFunction*/ true)),
                     colonToken = parseExpectedToken(SyntaxKind.ColonToken),
                     nodeIsPresent(colonToken)
-                        ? parseAssignmentExpressionOrHigher()
+                        ? parseAssignmentExpressionOrHigher(/*disallowReturnTypeInArrowFunction*/ true)

```

Comment:
Is it intentional that you're being conservative in the false branch? If not, shouldn't this propagate an outer `disallowReturnTypeInArrowFunction`?

Reply:
The problem here is when the conditional expression is nested but flat, as in this expression:

```ts
a ? b ? c : (d) : e => f
```

The intent is that it's parsed like this:

```ts
a ? (b ? c : (d)) : e => f
```

When parsing left to right, we'll be parsing the "false" side, and our scanner will contain:

```ts
(d) : e => f
```

If we aren't careful, we'll eat the colon because the caller said it was okay to parse it as an arrow expression with a return type.

But, I do think there's an improvement here; I should just inherit `allowReturnTypeInArrowFunction` from the caller. For most calls, it'll be `true` and allow the return type. Otherwise, we'll just remember downward that somewhere above, we're on the true side of a conditional expression.

Reply:
I can't however figure out a test case that this matters for, but I think modifying it is correct.

Reply:
FWIW the snippet `a ? b ? c : (d) : e => f` does not parse in today's TS, so I'm not worried about changing it.

Reply:
Well it's something like this that breaks in the "conservative" case, right?

```ts
a ? b : (c): Type => c
```

Reply:
Add the above as `parserArrowFunctionExpression17.ts`.